ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java

```diff
@@ -190,11 +240,12 @@ private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> s
         // If copy fails, fall through the retry logic
         LOG.info("file operation failed", e);
 
-        if (repeat >= (MAX_IO_RETRY - 1)) {
-          //no need to wait in the last iteration
+        if (repeat >= (MAX_IO_RETRY - 1)
+                || e.getMessage().equals(ErrorMsg.REPL_FILE_SYSTEM_OPERATION_RETRY.format(e.getCause().getMessage()))) {
+          //Don't retry if this is last iteration or retry is already exhausted by FS operations.
           break;
         }
-
+        closeAllForUGI((proxyUser == null) ? Utils.getUGI() : proxyUser);
```

Comment:
Now we closed the filesystem here. Below we recreate the filesystem only if exception is not `FNF` What will happen in case of a ``FileNotFoundException`` the filesystem will get closed but won't get recreated?